File: src/main/java/main/commands/prefixcommands/audio/SeekCommand.java
package main.commands.prefixcommands.audio;

import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;
import main.audiohandlers.RobertifyAudioManager;
import main.commands.prefixcommands.CommandContext;
import main.commands.prefixcommands.ICommand;
import main.utils.RobertifyEmbedUtils;
import main.utils.json.logs.LogType;
import main.utils.json.logs.LogUtils;
import main.utils.locale.RobertifyLocaleMessage;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.GuildVoiceState;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.internal.utils.tuple.Pair;

import javax.script.ScriptException;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.regex.Pattern;

@Deprecated
public class SeekCommand implements ICommand {
    @Override
    public void handle(CommandContext ctx) throws ScriptException {
        final List<String> args = ctx.getArgs();
        final Message msg = ctx.getMessage();
        final var guild = ctx.getGuild();

        if (args.isEmpty()) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "You must provide the position you would like to jump to").build())
                    .queue();
            return;
        }

        var position = ctx.getArgs().get(0);

        if (Pattern.matches("^\\d{1,2}:\\d{2}$", position)) {
            String[] positionSplit = position.split(":");
            int mins = Integer.parseInt(positionSplit[0]);
            int secs = Integer.parseInt(positionSplit[1]);

            final GuildVoiceState memberVoiceState = ctx.getMember().getVoiceState();
            final GuildVoiceState selfVoiceState = ctx.getSelfMember().getVoiceState();

            msg.replyEmbeds(handleSeek(selfVoiceState, memberVoiceState, mins, secs).build())
                    .queue();
        } else if (Pattern.matches("^\\d{1,2}:\\d{1,2}:\\d{2}$", position)) {
            String[] positionSplit = position.split(":");
            int hours = Integer.parseInt(positionSplit[0]);
            int mins = Integer.parseInt(positionSplit[1]);
            int secs = Integer.parseInt(positionSplit[2]);

            final GuildVoiceState memberVoiceState = ctx.getMember().getVoiceState();
            final GuildVoiceState selfVoiceState = ctx.getSelfMember().getVoiceState();

            msg.replyEmbeds(handleSeek(selfVoiceState, memberVoiceState, hours, mins, secs).build())
                    .queue();
        } else {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "You must provide the position in the format `mm:ss` **OR** `hh:mm:ss`").build())
                    .queue();
            return;
        }
    }

    public EmbedBuilder handleSeek(GuildVoiceState selfVoiceState, GuildVoiceState memberVoiceState, int mins, int sec) {
        final var guild = selfVoiceState.getGuild();

        if (!selfVoiceState.inVoiceChannel())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.VOICE_CHANNEL_NEEDED);

        if (!memberVoiceState.inVoiceChannel())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL_LOC, Pair.of("{channel}", selfVoiceState.getChannel().getAsMention()));

        if (memberVoiceState.getChannel().getIdLong() != selfVoiceState.getChannel().getIdLong())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL_LOC, Pair.of("{channel}", selfVoiceState.getChannel().getAsMention()));

        final var musicManager = RobertifyAudioManager.getInstance().getMusicManager(selfVoiceState.getGuild());
        final var audioPlayer = musicManager.getPlayer();

        if (audioPlayer.getPlayingTrack() == null)
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.NOTHING_PLAYING);

        if (mins < 0 || mins > 59)
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SeekMessages.INVALID_MINUTES);


        if (sec < 0 || sec > 59)
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SeekMessages.INVALID_SECONDS);


        long totalDurationInMillis = TimeUnit.MINUTES.toMillis(mins) + TimeUnit.SECONDS.toMillis(sec);

        AudioTrackInfo info = audioPlayer.getPlayingTrack().getInfo();
        if (totalDurationInMillis > info.length)
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SeekMessages.POS_GT_DURATION);

        audioPlayer.seekTo(totalDurationInMillis);
        final var time = (mins > 9 ? mins : "0" + mins) + ":" + (sec > 9 ? sec : "0" + sec);
        new LogUtils().sendLog(guild, LogType.TRACK_SEEK, RobertifyLocaleMessage.SeekMessages.SOUGHT_LOG,
                Pair.of("{user}", memberVoiceState.getMember().getAsMention()),
                Pair.of("{time}", time),
                Pair.of("{title}", info.title),
                Pair.of("{author}", info.author)
        );
        return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SeekMessages.SOUGHT,
                Pair.of("{time}", time),
                Pair.of("{title}", info.title),
                Pair.of("{author}", info.author)
        );

    }

    public EmbedBuilder handleSeek(GuildVoiceState selfVoiceState, GuildVoiceState memberVoiceState, int hours, int mins, int sec) {
        final var guild = selfVoiceState.getGuild();

        if (!selfVoiceState.inVoiceChannel())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.VOICE_CHANNEL_NEEDED);

        if (!memberVoiceState.inVoiceChannel())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL_LOC, Pair.of("{channel}", selfVoiceState.getChannel().getAsMention()));

        if (memberVoiceState.getChannel().getIdLong() != selfVoiceState.getChannel().getIdLong())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL_LOC, Pair.of("{channel}", selfVoiceState.getChannel().getAsMention()));

        final var musicManager = RobertifyAudioManager.getInstance().getMusicManager(selfVoiceState.getGuild());
        final var audioPlayer = musicManager.getPlayer();

        if (audioPlayer.getPlayingTrack() == null)
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.NOTHING_PLAYING);

        if (mins < 0 || mins > 59)
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SeekMessages.INVALID_MINUTES);


        if (sec < 0 || sec > 59)
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SeekMessages.INVALID_SECONDS);


        long totalDurationInMillis = TimeUnit.HOURS.toMillis(hours) + TimeUnit.MINUTES.toMillis(mins) + TimeUnit.SECONDS.toMillis(sec);

        AudioTrackInfo info = audioPlayer.getPlayingTrack().getInfo();
        if (totalDurationInMillis > info.length)
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SeekMessages.POS_GT_DURATION);

        audioPlayer.seekTo(totalDurationInMillis);
        final var time = (hours > 9 ? hours : "0" + hours) + ":"+ (mins > 9 ? mins : "0" + mins) + ":"+ (sec > 9 ? sec : "0" + sec);
        new LogUtils().sendLog(guild, LogType.TRACK_SEEK, RobertifyLocaleMessage.SeekMessages.SOUGHT_LOG,
                Pair.of("{user}", memberVoiceState.getMember().getAsMention()),
                Pair.of("{time}", time),
                Pair.of("{title}", info.title),
                Pair.of("{author}", info.author)
        );
        return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SeekMessages.SOUGHT,
                Pair.of("{time}", time),
                Pair.of("{title}", info.title),
                Pair.of("{author}", info.author)
        );
    }

    @Override
    public String getName() {
        return "seek";
    }

    @Override
    public String getHelp(String prefix) {
        return "Jump to a specific position in the current song\n" +
                "\nUsage: `"+ prefix+"seek <mm:ss>`\n";
    }

    @Override
    public boolean isPremiumCommand() {
        return true;
    }
}

File: src/main/kotlin/main/commands/slashcommands/audio/SeekCommand.kt
package main.commands.slashcommands.audio

import main.audiohandlers.RobertifyAudioManager
import main.audiohandlers.utils.author
import main.audiohandlers.utils.length
import main.audiohandlers.utils.title
import main.commands.slashcommands.SlashCommandManager.getRequiredOption
import main.utils.GeneralUtils.isNotNull
import main.utils.RobertifyEmbedUtils
import main.utils.RobertifyEmbedUtils.Companion.sendEmbed
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.SlashCommand
import main.utils.component.interactions.slashcommand.models.CommandOption
import main.utils.json.logs.LogType
import main.utils.json.logs.LogUtilsKt
import main.utils.locale.messages.SeekMessages
import net.dv8tion.jda.api.entities.GuildVoiceState
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.OptionType
import kotlin.time.DurationUnit
import kotlin.time.toDuration

class SeekCommand : AbstractSlashCommand(
    SlashCommand(
        name = "seek",
        description = "Jump to a specific position in the current song",
        options = listOf(
            CommandOption(
                type = OptionType.INTEGER,
                name = "hours",
                description = "The amount of hours to seek"
            ),
            CommandOption(
                type = OptionType.INTEGER,
                name = "minutes",
                description = "The amount of minutes to seek"
            ),
            CommandOption(
                type = OptionType.INTEGER,
                name = "seconds",
                description = "The amount of seconds to seek"
            ),
        )
    )
) {

    override suspend fun handle(event: SlashCommandInteractionEvent) {
        event.deferReply().queue()
        event.hook.sendEmbed {
            handleSeek(
                memberVoiceState = event.member!!.voiceState!!,
                selfVoiceState = event.guild!!.selfMember.voiceState!!,
                hours = event.getRequiredOption("hours").asInt,
                minutes = event.getRequiredOption("minutes").asInt,
                seconds = event.getRequiredOption("seconds").asInt
            )
        }.queue()
    }

    private fun handleSeek(
        memberVoiceState: GuildVoiceState,
        selfVoiceState: GuildVoiceState,
        hours: Int,
        minutes: Int,
        seconds: Int
    ): MessageEmbed {
        val guild = selfVoiceState.guild
        val acChecks = audioChannelChecks(memberVoiceState, selfVoiceState, songMustBePlaying = true)
        if (acChecks.isNotNull()) return acChecks!!

        if (minutes < 0 || minutes > 59)
            return RobertifyEmbedUtils.embedMessage(guild, SeekMessages.INVALID_MINUTES)
                .build()

        if (seconds < 0 || seconds > 59)
            return RobertifyEmbedUtils.embedMessage(guild, SeekMessages.INVALID_SECONDS)
                .build()

        val seekDuration =
            hours.toDuration(DurationUnit.HOURS).inWholeMilliseconds + minutes.toDuration(DurationUnit.MINUTES).inWholeMilliseconds + seconds.toDuration(
                DurationUnit.SECONDS
            ).inWholeMilliseconds

        val player = RobertifyAudioManager[guild].player
        val playingTrack = player.playingTrack

        if (seekDuration > playingTrack.length)
            return RobertifyEmbedUtils.embedMessage(guild, SeekMessages.POS_GT_DURATION)
                .build()

        player.seekTo(seekDuration)
        val time =
            "${if (hours > 9) "$hours" else "0$hours"}:${if (minutes > 9) "$minutes" else "0$minutes"}:${if (seconds > 9) "$seconds" else "0$seconds"}"

        LogUtilsKt(guild).sendLog(
            LogType.TRACK_SEEK,
            SeekMessages.SOUGHT_LOG,
            Pair("{user}", memberVoiceState.member.asMention),
            Pair("{time}", time),
            Pair("{title}", playingTrack.title),
            Pair("{author}", playingTrack.author)
        )
        return RobertifyEmbedUtils.embedMessage(
            guild,
            SeekMessages.SOUGHT,
            Pair("{time}", time),
            Pair("{title}", playingTrack.title),
            Pair("{author}", playingTrack.author)
        ).build()
    }

    override val help: String
        get() = """
                Jump to a specific position in the current song

                Usage: `/seek <hours> <minutes> <seconds>`
                """
}